File: server-pedidos-java/src/main/java/br/edu/utfpr/serverpedidosjava/controller/dto/EnderecoDTO.java
package br.edu.utfpr.serverpedidosjava.controller.dto;

import br.edu.utfpr.serverpedidosjava.entity.Cliente;
import br.edu.utfpr.serverpedidosjava.entity.Endereco;
import br.edu.utfpr.serverpedidosjava.util.validator.CEP;
import jakarta.validation.constraints.NotBlank;

public record EnderecoDTO(
        @NotBlank(message = "{cep.notblank}") @CEP String cep,
        @NotBlank(message = "{logradouro.notblank}") String logradouro,
        Integer numero,
        String complemento,
        String bairro,
        @NotBlank(message = "{cidade.notblank}") String cidade) {

    public Endereco toEntity(Cliente cliente) {
        Endereco endereco = new Endereco();
        endereco.setId(cliente.getId());
        endereco.setCep(this.cep);
        endereco.setLogradouro(this.logradouro);
        endereco.setNumero(this.numero);
        endereco.setComplemento(this.complemento);
        endereco.setBairro(this.bairro);
        endereco.setCidade(this.cidade);
        endereco.setCliente(cliente);
        return endereco;
    }

    public static EnderecoDTO fromEntity(Endereco endereco) {
        return new EnderecoDTO(
                endereco.getCep(),
                endereco.getLogradouro(),
                endereco.getNumero(),
                endereco.getComplemento(),
                endereco.getBairro(),
                endereco.getCidade());
    }
}

File: server-pedidos/src/main/kotlin/br/edu/utfpr/serverpedidos/controller/dto/EnderecoDTO.kt
package br.edu.utfpr.serverpedidos.controller.dto

import br.edu.utfpr.serverpedidos.entity.Cliente
import br.edu.utfpr.serverpedidos.entity.Endereco
import br.edu.utfpr.serverpedidos.util.validator.CEP
import jakarta.validation.constraints.NotBlank

data class EnderecoDTO(
    @field:NotBlank(message = "{cep.notblank}")
    @field:CEP
    val cep: String = "",
    @field:NotBlank(message = "{logradouro.notblank}")
    val logradouro: String = "",
    val numero: Int = 0,
    val complemento: String = "",
    val bairro: String = "",
    @field:NotBlank(message = "{cidade.notblank}")
    val cidade: String = ""
) {
    fun toEntity(cliente: Cliente): Endereco = Endereco(
        id = cliente.id,
        cep = this.cep,
        logradouro = this.logradouro,
        numero = this.numero,
        complemento = this.complemento,
        bairro = this.bairro,
        cidade = this.cidade,
        cliente = cliente
    )

    companion object {
        fun fromEntity(endereco: Endereco): EnderecoDTO = EnderecoDTO(
            cep = endereco.cep,
            logradouro = endereco.logradouro,
            numero = endereco.numero,
            complemento = endereco.complemento,
            bairro = endereco.bairro,
            cidade = endereco.cidade
        )
    }
}
